documentation/ArrayExpressionLoweringToFIR.md

```diff
@@ -0,0 +1,315 @@
+    jmp bresult(%6)
+  ^bnullshape:
+    // Create a null shaped result.
+    %nullres = undef !fir.array<0:f32>
+    %r = fir.convert %nullres !fir.array<?:f32> // not sure we need this
+    jmp bresult(%r)
+  ^bresult(%result : !fir.array<?:f32>)
+    //%result is the expression result
+
+  // The assignment is not part of the expression lowering
+  // It is also more complex because y is an allocatable so not shown
+  // here
+```
+
+Observation: The loop is on the shape, not the actual bounds of `x`. So we need to compute the related coordinate for `x` inside the loop.
+Here, it is a bit stupid, but in `x + y`, it does not make sense to loop on `x` bounds because the bounds of y might be different, so we loop on
+the extent starting from 1.
+
+Questions: Does it matter? Could we hide this inside `fir.coordinate_of` ?
+
+### Short-circuiting example
+Scalar driven short-circuiting
+In case `b1` is scalar and `foo()` has side effects in `b1.AND.foo()`,
+f18 guarantees `foo()` will only be called if required (when `b1` is true).
+This short circuiting logic has to be evaluated before the `fir.loop`.
+
+Fortran:
+```
+ real :: x(4), y(4), z(4)
+ z = merge(x, y, b1.AND.foo())
+```
+`foo` returns an array with `x` shape. 
+
+FIR:
+```
+  ^bb0
+    %tmp = fir.alloca !fir.array<4:f32>
+    %1 = fir.load b1
+    %2 = fir.convert b1 i1
+    %3 = cond_br %i1 ^bb1 ^bb2(%1)
+  ^bb1:
+    %4 = fir.call "foo"
+    jmp ^bb3(%4)
+  ^bb2
+    %tmp2 = fir.alloca !fir.array<4:fir.logical<4>>
+    for.loop j 1 to 4 {
+      %5 = fir.coordinate_of %tmp, %i
+      fir.store %1 to %5
+    }
+    %6 = fir.load %tmp2
+    jmp ^bb3(%6)
+  ^bb3(%bool : fir.array<4:fir.logical<4>>)
+    fir.loop %i = 1 to 4 {
+    %7 = fir.extract_value %x, %i
+    %8 = fir.extract_value %y, %i
+    %9 = fir.call "fir.merge.i32" %7 %8 %bool
+    %10 = fir.coordinate_of %tmp, %i
+    fir.store %9 to %10
+    }
+    %11 = fir.load %tmp // This is the expression result
+
+    // The assignment part is not part of the expression lowering
+    fir.store %11 to %z
+```
+
+
+Observation: In the example above, in case `b1` is false, we are going through two loops for nothing.
+(the result is simply `fir.load %y`). We could instead have rewritten the expression as
+`merge(merge(x, y, foo()), y, b1)` which would lead to FIR:
+```
+  ^bb0
+    %1 = fir.load b1
+    %2 = fir.convert b1 i1
+    %3 = cond_br %i1 ^bb1 ^bb2(%y)
+  ^bb1:
+    %tmp = fir.alloca !fir.array<4:f32> 
+    %4 = fir.call "foo"
+    fir.loop %i = 1 to 4 {
+      %5 = fir.extract_value %x, %i
+      %6 = fir.extract_value %y, %i
+      %7 = fir.extract_value %4, %i
+      %8 = fir.call "fir.merge.i32" %5 %6 %7
+      %9 = fir.coordinate_of %tmp, %i
+      fir.store %8 to %9
+    }
+    jmp ^bb3(%4)
+  ^bb2(%resRef : fir.ref<fir.array<4:f32>>)
+    %res = fir.load %resRef // This is the expression result
+  ```
+
+Question: Can optimizers still do this kind of rewriting later?
```

Comment:
Does `MERGE` do anything special beyond selecting one or the other value?

It seems like we want to lower `MERGE` using an algorithm something like:

1) Does this expression have a conditional guard and does the guard short-circuit?
    yes - generate the apropos `br cond` test
2) Generate a `loop` turn the guard into a `where`

Reply:
OK, so we should not use `fir.where` to implement if/else short-circuits but prefer explicit CFG with `br cond` ?  I thought because we could use `fir.where` outside of `fir.loop` they could be used to implement `if/else` logic in expression lowering.

Reply:
Either will work.